[SR-153] Add IsClosureCapture flag to VarDecl

gregomni opened the pull request:
That flag helps distinguish between truly constant vars and those which
aren’t assignable but need to be mutable (like weak closure captures).
Also fixes sr-153 by improving error reporting in these cases.

gregomni commented:
Tests added and all tests pass. Should this flag be added to the ASTDumper as well?

slavapestov commented:
Looks reasonable at first glance, @lattner ?

DougGregor commented:
"IsUserAssignable" seems like the wrong name for this behavior. Let's state what the variable actually is---a capture---and derive the assignability behavior from that. Also, we either need (de-)serialization code or we should assert() that a capture variable never gets serialized (I believe the latter should be the case), so we don't get surprised later on if capture variables start to get serialized.

gregomni commented:
@DougGregor I can certainly do that. I asked @lattner about this bug on swift-dev, and he implied that there might be other places where this flag could be useful (in fact he suggested perhaps being even more general and changing isLet into an enum for assignable { zero, once, many } to correspond to captures / lets / vars, and then an entirely separate flag for constness as far as the rest of the compiler is concerned). 

I couldn't actually any other places in Swift at the moment besides weak captures where isUserAssignable would be false, but I could certainly imagine it being nice in the future to be able to declare weak lets, that the compiler knows is !isLet (in the it-can-change-on-you sense) but is also !isUserAssignable (compiler errors if user code assigns to it).

Anyway, that's the thought behind the more general name for the behavior, but maybe to make the generality make sense I would've needed to do the much larger semantic change to all VarDecls, and since I chickened out and made the tiny change, a very specific isCapture name would be more appropriate. :-)

gregomni commented:
@DougGregor This seemed to be languishing so I went ahead and rebased off master, changed over to use the more specific IsClosureCapture for the flag name, and asserted that they don't get serialized.

Unsquashed at the moment so you can see the changes in context, but if this looks good, let me know and I'll squash down.
